File: data/sorting/counting-sort.py
def countingSort(nums: list[int]) -> list[int]:
    """
    Counting Sort Algorithm
    Time Complexity: O(n + k)
    Space Complexity: O(n + k)
    """
    if not nums:
        return nums

    min_val, max_val = min(nums), max(nums)
    count = [0] * (max_val - min_val + 1)

    for num in nums:
        count[num - min_val] += 1

    sorted_idx = 0
    for i, freq in enumerate(count):
        val = i + min_val
        for _ in range(freq):
            nums[sorted_idx] = val
            sorted_idx += 1

    return nums

File: data/sorting/counting-sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> countingSort(vector<int>& nums) {
    if (nums.empty()) return nums;

    int minVal = *min_element(nums.begin(), nums.end());
    int maxVal = *max_element(nums.begin(), nums.end());
    vector<int> count(maxVal - minVal + 1, 0);

    for (int num : nums) {
        count[num - minVal]++;
    }

    int idx = 0;
    for (int i = 0; i < count.size(); ++i) {
        while (count[i] > 0) {
            nums[idx++] = i + minVal;
            count[i]--;
        }
    }
    return nums;
}
